File: sandbox-runner/patterns/fixed-count-retries/typescript/starter.ts
import { Client, Connection } from "@temporalio/client";

import { TASK_QUEUE, WORKFLOW_ID_PREFIX } from "./shared";
import { paymentWorkflow } from "./workflows";

async function main(): Promise<void> {
  const connection = await Connection.connect();
  try {
    const client = new Client({ connection });
    const orderId = `order-${Date.now()}`;

    console.log(`Starting payment workflow for ${orderId}`);
    console.log("The payment gateway is down — all 3 attempts will fail.");
    console.log();

    const handle = await client.workflow.start(paymentWorkflow, {
      args: [orderId],
      taskQueue: TASK_QUEUE,
      workflowId: `${WORKFLOW_ID_PREFIX}-${orderId}`,
    });
    console.log(`Started workflow: ${handle.workflowId}`);

    const result = await handle.result();
    console.log(`Workflow result: ${result}`);
    console.log();
    console.log(
      `Open the Temporal UI and search for '${handle.workflowId}' to inspect the retry history.`,
    );
  } finally {
    await connection.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: sandbox-runner/patterns/fixed-count-retries/python/starter.py
import asyncio
import time

from temporalio.client import Client

from shared import TASK_QUEUE, WORKFLOW_ID_PREFIX
from workflows import PaymentWorkflow


async def main() -> None:
    client = await Client.connect("localhost:7233")
    order_id = f"order-{int(time.time() * 1000)}"

    print(f"Starting payment workflow for {order_id}")
    print("The payment gateway is down — all 3 attempts will fail.")
    print()

    handle = await client.start_workflow(
        PaymentWorkflow.run,
        order_id,
        id=f"{WORKFLOW_ID_PREFIX}-{order_id}",
        task_queue=TASK_QUEUE,
    )

    result = await handle.result()
    print(f"Workflow result: {result}")
    print()
    print(f"Open the Temporal UI and search for '{handle.id}' to inspect the retry history.")


if __name__ == "__main__":
    asyncio.run(main())
